streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2469,6 +2469,60 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
       } yield pull
     }
 
+  final def debounce(d: Duration): ZStream[R with Clock, E, O] = {
+    type Loser = Either[Fiber[Nothing, Take[E, O]], Fiber[Nothing, Take[E, O]]]
+
+    def race(previous: URIO[R, Take[E, O]], current: URIO[R, Take[E, O]]): URIO[R, (Take[E, O], Loser)] =
+      previous.raceWith(current)(
+        (exit, current) => ZIO.done(exit).map(a => (a, Right(current))),
+        (exit, previous) => ZIO.done(exit).map(b => (b, Left(previous)))
+      )
+
+    def store(ref: Ref[Option[Fiber[E, Chunk[O]]]])(chunk: Chunk[O]): URIO[Clock, Chunk[O]] =
+      for {
+        fiber <- clock.sleep(d).as(chunk).fork
+        _     <- ref.set(Some(fiber))
+        empty <- Pull.emit(Chunk.empty)
+      } yield empty
+
+    ZStream[R with Clock, E, O] {
+      for {
+        chunks <- self.process
+        ref    <- Ref.make[Option[Fiber[E, Chunk[O]]]](None).toManaged_
+        pull = {
+          ref.get.flatMap {
+            case None =>
+              chunks >>= store(ref)
+            case Some(previous) =>
+              val prev = previous.join.run.map(_.mapError(Some(_)))
+
+              race(prev, chunks.run).flatMap {
+                case (exit, loser) =>
+                  exit.foldM(
+                    Cause.sequenceCauseOption(_) match {
+                      case Some(e) =>
+                        loser.merge.interrupt *> Pull.halt(e)
+                      case None =>
+                        for {
+                          fiber  <- loser.merge.join
+                          done   <- ZIO.done(fiber)
+                          _      <- ref.set(None)
+                          result <- Pull.emit(done)
+                        } yield result
+                    },
+                    chunk =>
+                      loser.fold(
+                        _.interrupt *> store(ref)(chunk),
+                        _.interrupt *> ref.set(None) *> Pull.emit(chunk)
```

Comment:
We can't interrupt an ongoing pull. That is likely to lead to corruption of internal operator state. We have to keep the ongoing pull around.

Reply:
@iravid solved